### PR TITLE
Issue 1282: base64_decode, base64_data - v7

### PR DESCRIPTION
**Pull request by jasonish**
Fixes issues in last PR: https://github.com/inliniac/suricata/pull/1690

For util-base64, intoduce "strict" mode so other uses keep the same behaviour.

Redmine ticket: https://redmine.openinfosecfoundation.org/issues/1282

Buildbot output:
- PR jasonish-pcap: https://buildbot.openinfosecfoundation.org/builders/jasonish-pcap/builds/154
- PR jasonish: https://buildbot.openinfosecfoundation.org/builders/jasonish/builds/155

### CODE REVIEW

File: src/detect-base64-decode.c [hunk 1]
```diff
@@ -0,0 +1,733 @@
...
+        printf("Decoded data:\n");
+        PrintRawDataFp(stdout, det_ctx->base64_decoded,
+            det_ctx->base64_decoded_len);
+    }
+#endif
+
+    return det_ctx->base64_decoded_len > 0;
+}
+
+static int ParseUint32(const char *str, uint32_t *value)
```
Comment (inliniac) [inline src/detect-base64-decode.c]: We can use ByteExtractUint32 instead?

File: src/detect-engine-content-inspection.c [hunk 2]
```diff
@@ -551,6 +553,16 @@ int DetectEngineContentInspection(DetectEngineCtx *de_ctx, DetectEngineThreadCtx
         SCLogDebug("lua match");
         goto match;
 #endif /* HAVE_LUA */
+    } else if (sm->type == DETECT_BASE64_DECODE) {
+        if (DetectBase64DecodeDoMatch(det_ctx, s, sm, buffer, buffer_len)) {
+            if (s->sm_arrays[DETECT_SM_LIST_BASE64_DATA] != NULL) {
+                KEYWORD_PROFILING_END(det_ctx, sm->type, 1);
```
Comment (inliniac) [inline src/detect-engine-content-inspection.c]: Seems this is called elsewhere regardless of the result here. Do we need it?

File: src/detect-base64-decode.c (hunk 1 above)
Comment (jasonish) [inline src/detect-base64-decode.c]: > We can use ByteExtractUint32 instead?

Isn't that for extracting a uint32 from packed bytes? ParseUint32 is to parse an integer from text. A lot of other code uses atoi, but I'm not a fan of that function. Perhaps we need some generic integer parsing functions so we can do away with atoi?

File: src/detect-engine-content-inspection.c (hunk 2 above)
Comment (jasonish) [inline src/detect-engine-content-inspection.c]: Yes, we need it before the call to DetectBase64DataDoMatch as that will reenter DetectEngineContentInspection with profiling on, causing Suricata to bail at line 110 in this file.

File: src/detect-base64-decode.c (hunk 1 above)
Comment (inliniac) [inline src/detect-base64-decode.c]: ByteExtractStringUint32 is for strings, so perhaps that? Would need to consider how errors are handled.

File: src/detect-engine-content-inspection.c (hunk 2 above)
Comment (inliniac) [inline src/detect-engine-content-inspection.c]: ack

File: src/detect-base64-decode.c (hunk 1 above)
Comment (jasonish) [inline src/detect-base64-decode.c]: That works almost as a drop in here. prscript running now.

**Comment by jasonish** [comment]
Closed. See #1746.

**Pull request closed without merging**